### PR TITLE
fix(es): add java date format support

### PR DESCRIPTION
**Pull request by etolbakov**
### Description
add conversion from elasticsearch/opensearch date format to quickwit expected strptime

### How was this PR tested?
Provide curl output below 

### Related Issue
https://github.com/quickwit-oss/quickwit/issues/5228

The code is take from https://github.com/quickwit-oss/quickwit/pull/5237

### CODE REVIEW

File: quickwit/quickwit-datetime/src/date_time_format.rs [hunk 1]
```diff
@@ -17,46 +17,102 @@
...
+            (r#"'[^']+'"#, |s| {
+                Some(literal(s[1..s.len() - 1].as_bytes()))
+            }),
+            (r#"[^\w\[\]{}]"#, |s| Some(literal(s.as_bytes()))),
+        ])
+        .unwrap()
+    })
+}
+
+fn java_date_format_aliases() -> &'static HashMap<&'static str, &'static str> {
```
Comment (etolbakov) [inline quickwit/quickwit-datetime/src/date_time_format.rs:94]: will add more mappings if we decide to proceed/support it

**Comment by etolbakov** [comment]
@fulmicoton just a ping about this change.
Could you please take a look when you have time?

File: quickwit/quickwit-datetime/src/date_time_format.rs [hunk 2]
```diff
@@ -104,20 +160,49 @@ impl StrptimeParser {
...
+    }
+
+    pub fn from_java_datetime_format(java_datetime_format: &str) -> Result<StrptimeParser, String> {
+        let format = if java_datetime_format.contains('_') {
+            java_date_format_aliases()
+                .get(java_datetime_format)
+                .unwrap_or(&java_datetime_format)
+        } else {
+            java_datetime_format
+        };
```
Comment (fulmicoton) [inline quickwit/quickwit-datetime/src/date_time_format.rs:186]:
```suggestion
        // We check if the date format is present in our alias dictionary.
        let java_date_format_resolved: &str = java_date_format_aliases()
                .get(java_datetime_format)
                .unwrap_or(&java_datetime_format);
```

File: quickwit/quickwit-query/Cargo.toml [hunk 3]
```diff
@@ -18,10 +18,12 @@ lindera-core = { workspace = true, optional = true }
 lindera-dictionary = { workspace = true, optional = true }
 lindera-tokenizer = { workspace = true, optional = true }
 once_cell = { workspace = true }
+regex = { workspace = true }
```
Comment (fulmicoton) [inline quickwit/quickwit-query/Cargo.toml:21]:
```suggestion
```

File: quickwit/quickwit-query/Cargo.toml [hunk 4]
```diff
@@ -18,10 +18,12 @@ lindera-core = { workspace = true, optional = true }
...
 serde = { workspace = true }
 serde_json = { workspace = true }
 serde_with = { workspace = true }
 tantivy = { workspace = true }
+time = { workspace = true }
```
Comment (fulmicoton) [inline quickwit/quickwit-query/Cargo.toml:26]:
```suggestion
```

File: quickwit/quickwit-datetime/src/date_time_format.rs [hunk 5]
```diff
@@ -17,46 +17,102 @@
...
+// Elasticsearch/OpenSearch uses a set of preconfigured formats, more information could be found
+// here https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html
+fn java_date_format_tokenizer() -> &'static RegexTokenizer<OwnedFormatItem> {
+    static JAVA_DATE_FORMAT_TOKENIZER: OnceLock<RegexTokenizer<OwnedFormatItem>> = OnceLock::new();
+    JAVA_DATE_FORMAT_TOKENIZER.get_or_init(|| {
+        RegexTokenizer::new(vec![
+            (r#"yy(yy)?"#, build_year_item),
+            (r#"MM?"#, build_month_item),
+            (r#"dd?"#, build_day_item),
+            (r#"''"#, |_| Some(literal(b"'"))),
```
Comment (fulmicoton) [inline quickwit/quickwit-datetime/src/date_time_format.rs:84]: we are still missing a lot of the patterns... Hours etc.

**Comment by fulmicoton** [comment]
We need rest unit tests and to handle more patterns.

File: quickwit/quickwit-datetime/src/date_time_format.rs [hunk 6]
```diff
@@ -17,46 +17,102 @@
...
+    })
+}
+
+fn java_date_format_aliases() -> &'static HashMap<&'static str, &'static str> {
+    static JAVA_DATE_FORMAT_ALIASES: OnceLock<HashMap<&'static str, &'static str>> =
+        OnceLock::new();
+    JAVA_DATE_FORMAT_ALIASES.get_or_init(|| {
+        let mut m = HashMap::new();
+        m.insert("date_optional_time", "yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+        m.insert("strict_date_optional_time", "yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
Comment (fulmicoton) [inline quickwit/quickwit-datetime/src/date_time_format.rs:100]: I don't think we handle this format right now.

File: quickwit/quickwit-query/Cargo.toml (hunk 4 above)
Comment (etolbakov) [reply quickwit/quickwit-query/Cargo.toml:25]: This one is required by `range_query.rs`
```
use time::format_description::well_known::Rfc3339;
```

**Comment by kuzaxak** [comment]
@etolbakov Can I reopen it from local branch in QW repo?

**Comment by kuzaxak** [comment]
@etolbakov I think we can close this one and continue in a local branch [here](https://github.com/quickwit-oss/quickwit/pull/5462) I refactored you tokenizer to support nested optional and uncommented test cases (they are passing on my local).

**Pull request closed without merging**